Public/Src/Cache/ContentStore/Distributed/Sessions/DistributedContentSession.cs

```diff
@@ -165,5 +181,32 @@ private async Task<PutResult> RegisterPutAsync(Context context, CancellationToke
 
             return putResult;
         }
+
+        private async Task<BoolResult> RequestProactiveCopyIfNeededAsync(OperationContext context, ContentHash hash, UrgencyHint urgencyHint)
+        {
+            var hashArray = new[] { hash };
+
+            // First check in local location store, then global if failed.
+            var getLocationsResult = await ContentLocationStore.GetBulkAsync(context, hashArray, context.Token, UrgencyHint.Nominal, GetBulkOrigin.Local);
+            if (!getLocationsResult.Succeeded || getLocationsResult.ContentHashesInfo[0].Locations.Count == 0)
+            {
+                getLocationsResult = await ContentLocationStore.GetBulkAsync(context, hashArray, context.Token, UrgencyHint.Nominal, GetBulkOrigin.Global);
+            }
+
+            if (getLocationsResult.Succeeded)
+            {
+                if (getLocationsResult.ContentHashesInfo[0].Locations.Count == 1)
+                {
+                    var locations = ContentLocationStore.GetKnownMachineLocations();
+                    var location = locations[ThreadSafeRandom.Generator.Next(locations.Length)];

```

Comment:
Deal with corner case that randomly chosen location is self. Suggest while (location == self) { get new location } so that we never exit with location == self.